uberfire-m2repo-editor/uberfire-m2repo-editor-backend/src/main/java/org/guvnor/m2repo/backend/server/FixNotMavenizedArtifactInfo.java

```diff
@@ -0,0 +1,90 @@
+package org.guvnor.m2repo.backend.server;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.Properties;
+import java.util.TreeMap;
+
+import org.guvnor.common.services.project.model.GAV;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.guvnor.m2repo.backend.server.M2ServletContextListener.ARTIFACT_ID;
+import static org.guvnor.m2repo.backend.server.M2ServletContextListener.GROUP_ID;
+import static org.guvnor.m2repo.backend.server.M2ServletContextListener.VERSION;
+
+/**
+ * This class aim to provide maven meta information for artifacts (.jar) that don't
+ * have maven metadata built-in.
+ * <p>
+ * Maven auto-generate a `pom.properties` to store GAV information, and it also
+ * embeds the original pom file into the produced artifacts. However, this meta information
+ * is not always available if the artifact was created using a different system like ant.
+ * Even if the artifact doesn't have the necessary embedded information, those artifacts can
+ * be available in maven repositories, but the metadata are provided differently.
+ * <p>
+ * To build the GAV information out of a .jar file, this class uses as input a
+ * `not-mavenized-artifacts.properties` from classpath that provides the feedback needed
+ * to produce the required metadata. In case of an artifact is not listed in the input file,
+ * this class won't be able to provide the GAV.
+ */
+public class FixNotMavenizedArtifactInfo {
+
+    private static final Logger logger = LoggerFactory.getLogger(FixNotMavenizedArtifactInfo.class);
+
+    private final TreeMap<String, String[]> notMavenizedArtifacts = new TreeMap<>((o1, o2) -> {
```

Comment:
It took me several minutes to understand this weird comparator. In the end I understood the intention, but without knowing the context, this might seem like a bug to sort long strings before shorter ones.

In this case comment and maybe more readable comparator using Java 8 apis would have been better. Also you don't need to do the final comparison of same-length strings as they wouldn't be equal anyway.

```java
    // Sort artifact names by length from longer to shorter to avoid false positives when shorter prefix matches too (e.g.  ant-launcher vs. ant)
    private static final Comparator<String> LONG_NAMES_BEFORE_SHORT_COMPARATOR = Comparator.comparing(String::length).reversed();
    private final TreeMap<String, String[]> notMavenizedArtifacts = new TreeMap<>(LONG_NAMES_BEFORE_SHORT_COMPARATOR);
```

Reply:
I could add the comment, and even make it a constant, but I don't think I could use the Java8 comparator, as the goal is not just sort by length, but also sort alphabetically

Reply:
Looking at how the map keys are used in this class (`fullFileName.startsWith(entry.getKey())`) the order of prefixes with the same lenght should not matter. You can leave it as is now, but know (for the future) that java 8 apis are quite powerful. In particular the comparator that you have could have been made a bit more readable like so:

```java
    private static final Comparator<String> LONG_NAMES_BEFORE_SHORT_COMPARATOR = Comparator
            .comparing(String::length).reversed()
            .thenComparing(naturalOrder());
```